### PR TITLE
Add support for macOS on Apple Silicon

### LINKED ISSUE #138: ld: library not found for -lsharpyuv
I've got the following error when trying to use lilliput 

```
/opt/homebrew/Cellar/go@1.20/1.20.10/libexec/pkg/tool/darwin_arm64/link: running c++ failed: exit status 1
ld: library not found for -lsharpyuv
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

### LINKED ISSUE #157: A path to supporting ARM
Hello, all. There seems to be demand for ARM support in Lilliput, and several folks have proposed how to do it, but it's not clear what the project maintainers would be willing to accept contribution-wise. My hope is to gather the relevant threads here and to start that conversation.

Most of the friction is related to supporting Apple Silicon so that developers can build Lilliput on modern Mac hardware. There's also friction, though perhaps less, related to building for arm64 on Linux to be deployed in production environments. There's a general expectation that packages support both x86_64 and arm64 these days.

I see three main problems to navigate:
1. Deciding which combinations of OS and architecture to support. A maintainer [has suggested](https://github.com/discord/lilliput/pull/131#issuecomment-1709035845) that the x86 deps for macOS might not be worth providing anymore. The author of #98 mentioned that including deps for `{darwin,linux}x{x86_64,arm64}` approaches the 500MB limit for Go modules. Supporting linux/x86_64 seems non-negotiable. What else should be included?
2. Updating the cgo directives in `{avcodec,giflib,opencv,thumbhash,webp}.go` to support multiple architectures. This is straightforward (example: #98) once we know the OSes and architectures that Lilliput should support.
3. It's not clear how the pre-compiled dependencies are actually built and packaged in Discord's CI system, which makes it difficult for community devs to build Lilliput for environments that aren't (yet) supported. It's also difficult to do a supply chain audit of the dependency build, which is increasingly important to many developers. While we don't need to resolve this issue in order to support ARM, I think it would be very helpful.

I've made a [prototype branch](https://github.com/discord/lilliput/compare/master...wistia:lilliput:macos-arm64-support) to make clear what I'm proposing, specifically [this commit](https://github.com/wistia/lilliput/commit/0ca04dfd4b1e981e894205b79d4fe36a5ec0df12) for the CI workflow and [this one](https://github.com/wistia/lilliput/commit/a28f85ceffd55290200e3e8d8f95b4ad650133dd) for adding support for Apple Silicon. The rest of the commits address various build issues that I ran into. There are sample CI runs for [building dependencies](https://github.com/wistia/lilliput/actions/runs/8992104172) and [building and testing the package](https://github.com/wistia/lilliput/actions/runs/8992104171).

### Questions for maintainers

1. Would you accept a PR that adds support for macOS on ARM and drops support for macOS on Intel? If not, what combination would you prefer to support?
2. Would you welcome a public-facing CI workflow that uses GitHub Actions?

@brian-armstrong-discord @skidder @walruscow (recent committers 👋)

Related: #95 #98 #111 #138 #139

### PR DESCRIPTION
**Pull request by acj**
As discussed in #157, this PR contributes the following:
1. Adds dependencies built for macOS on ARM / Apple Silicon
2. Removes dependencies built for macOS on Intel x86_64
3. Adds GitHub Actions CI workflows for building the dependencies and for building and testing Lilliput itself
4. Fixes several issues that were breaking the build on macOS and Linux

I tried to break things up into small commits for easier review. The last ~5 commits are the key changes -- adding arm64 support, adding CI, and rebuilding the dependencies. The rest are fixes for the build.

If you'd like me to make a separate PR for the CI stuff, I'm happy to do that. There's a test run of the main CI workflow [here](https://github.com/wistia/lilliput/actions/runs/9215316841) and one for the dependency workflow [here](https://github.com/wistia/lilliput/actions/runs/9215207089).

Happy to talk through any questions or concerns. Thanks!

Resolves #157. Resolves #138. Resolves #139.

### CODE REVIEW

**Comment by acj** [comment]
@skidder :wave: A gentle ping. If I can do anything to make this easier to review and incorporate, let me know.

**Comment by aslamhadi** [comment]
hi, any updates on this PR?

File: deps/build-deps-osx.sh [hunk 1]
```diff
@@ -24,6 +39,10 @@ rm -rf ffmpeg
 
 if [ ! -d "$SRCDIR" ]; then
     git clone https://github.com/discord/lilliput-dep-source "$SRCDIR"
+
+    # TODO: Fix this in the upstream deps repo
+    cd "$SRCDIR"
+    wget -q -O giflib-5.2.2.tar.gz https://sourceforge.net/projects/giflib/files/giflib-5.2.2.tar.gz/download
```
Comment (Waldeedle) [inline deps/build-deps-osx.sh:45]: seems like 5.2.2 has been added in the upstream deps repo already and that discord has added a patch for giflib as well

Comment (Waldeedle) [reply deps/build-deps-osx.sh:45]: @remiolivier-discord I noticed you updated the deps repo, could you also update the build scripts in the main branch for lilliput? 🙏

**Comment by skidder** [comment]
> @skidder Thanks for landing #192. 👍 Are you interested in the CI workflow from this PR? If so, I can rebase and remove the changes that are already in master.

@acj That would be amazing, yes, please!

**Comment by acj** [comment]
I opened #196 for the CI workflows.

**Pull request closed without merging**